libsplinter/src/admin/service/event/store/diesel/operations/list_events_by_management_type_since.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides the "list events by management type" operation for the `DieselAdminServiceEventStore`.
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use diesel::{prelude::*, types::HasSqlType};
+
+use super::AdminServiceEventStoreOperations;
+
+use crate::admin::service::event::{
+    store::{
+        diesel::{
+            models::{
+                AdminEventCircuitProposalModel, AdminEventProposedCircuitModel,
+                AdminEventProposedNodeModel, AdminEventProposedServiceArgumentModel,
+                AdminEventProposedServiceModel, AdminEventVoteRecordModel, AdminServiceEventModel,
+            },
+            schema::{
+                admin_event_circuit_proposal, admin_event_proposed_circuit,
+                admin_event_proposed_node, admin_event_proposed_node_endpoint,
+                admin_event_proposed_service, admin_event_proposed_service_argument,
+                admin_event_vote_record, admin_service_event,
+            },
+        },
+        AdminServiceEventStoreError, EventIter,
+    },
+    AdminServiceEvent,
+};
+use crate::admin::store::{
+    AuthorizationType, CircuitProposalBuilder, DurabilityType, PersistenceType, ProposalType,
+    ProposedCircuitBuilder, ProposedNode, ProposedNodeBuilder, ProposedService,
+    ProposedServiceBuilder, RouteType, VoteRecord,
+};
+
+pub(in crate::admin::service::event::store::diesel) trait AdminServiceEventStoreListEventsByManagementTypeSinceOperation
+{
+    fn list_events_by_management_type_since(
+        &self,
+        management_type: String,
+        start: i64,
+    ) -> Result<EventIter, AdminServiceEventStoreError>;
+}
+
+impl<'a, C> AdminServiceEventStoreListEventsByManagementTypeSinceOperation
+    for AdminServiceEventStoreOperations<'a, C>
+where
+    C: diesel::Connection,
+    C::Backend: HasSqlType<diesel::sql_types::BigInt>,
+    String: diesel::deserialize::FromSql<diesel::sql_types::Text, C::Backend>,
+    i64: diesel::deserialize::FromSql<diesel::sql_types::BigInt, C::Backend>,
+    Vec<u8>: diesel::deserialize::FromSql<diesel::sql_types::Binary, C::Backend>,
+{
+    fn list_events_by_management_type_since(
```

Comment:
Is this `list_by` operation only different from the other one due to a difference in the initial filter?

Reply:
Yeah, similar to how the `AdminServiceStore` lists things based on the circuit predicates, but the rest of the operation is relatively the same. It takes the event IDs from that filtered query, and uses those to filter the later queries.

Reply:
can these be consolidated into a single operation?

Reply:
Probably, one that only the event IDs are passed to. Assuming you mean within the separate `list_since` and `list_by_management_type_since` operations, and not altering the trait's operations (which could also probably be consolidated by just making the management type optional).

Reply:
Right, I'm only saying at the operation level.  The trait should remain as it is.

Reply:
Updated to include a `list_events` operation that just takes a list of event IDs and holds all of the code that translates the database objects to the native structs, used by both the `list_*` operations.